src/api/bulk2.ts

```diff
@@ -0,0 +1,1000 @@
+      ? getLogger('bulk2:QueryJobV2').createInstance(this.connection._logLevel)
+      : getLogger('bulk2:QueryJobV2');
+    this.body = options.body;
+    this.pollingOptions = options.pollingOptions;
+    // default error handler to keep the latest error
+    this.on('error', (error) => (this.error = error));
+  }
+
+  /**
+   * Creates a query job
+   *
+   * @returns {Promise<QueryJobInfoV2>} job information.
+   */
+  async open(): Promise<QueryJobInfoV2> {
+    try {
+      this.jobInfo = await this.createQueryRequest<QueryJobInfoV2>({
+        path: '',
+        method: 'POST',
+        body: JSON.stringify(this.body),
+        headers: {
+          'Content-Type': 'application/json; charset=utf-8',
+        },
+        responseType: 'application/json',
+      });
+      this.logger.debug(this.jobInfo);
+      this.emit('open');
+    } catch (err) {
+      this.emit('error', err);
+      throw err;
+    }
+
+    return this.jobInfo;
+  }
+
+  /**
+   * Abort the job
+   *
+   * The 'aborted' event is emitted when the job successfully aborts.
+   * @returns {Promise<QueryJobInfoV2>} job information.
+   */
+  async abort(): Promise<QueryJobInfoV2> {
+    try {
+      const state: JobStateV2 = 'Aborted';
+      this.jobInfo = await this.createQueryRequest<QueryJobInfoV2>({
+        method: 'PATCH',
+        path: `/${this.jobInfo?.id}`,
+        body: JSON.stringify({ state }),
+        headers: { 'Content-Type': 'application/json; charset=utf-8' },
+        responseType: 'application/json',
+      });
+      this.logger.debug(this.jobInfo);
+      return this.jobInfo;
+    } catch (err) {
+      this.emit('error', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Poll for the state of the processing for the job.
+   *
+   * @param interval Polling interval in milliseconds
+   * @param timeout Polling timeout in milliseconds
+   * @returns {Promise<Record[]>} A promise that resolves when the job finished being processed.
+   */
+  async poll(
+    interval: number = this.pollingOptions.pollInterval,
+    timeout: number = this.pollingOptions.pollTimeout,
+  ): Promise<void> {
+    const jobId = getJobId(this.jobInfo);
+    const startTime = Date.now();
+
+    this.logger.debug(`Job ID: ${jobId}`);
+    this.logger.debug(
+      `Polling options: timeout:${timeout}ms | interval: ${interval}ms.`,
+    );
+
+    while (startTime + timeout > Date.now()) {
+      try {
+        const res = await this.check();
+        switch (res.state) {
+          case 'Aborted':
+            throw new Error('Job has been aborted');
+          case 'UploadComplete':
+          case 'InProgress':
+            await delay(interval);
+            break;
+          case 'Failed':
+            // unlike ingest jobs, the API doesn't return an error msg:
+            // https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/query_get_one_job.htm
+            this.logger.debug(res);
+            throw new Error('Query job failed to complete');
```

Comment:
I put the response in debug logs but users will need to check in their org if three's more info about a query job failre.